fix: #115 : Infoview not working for external file

lovettchris opened the pull request:
Fixes #115

Turns out the work to support "untitled" files broke the "ad hoc" files.   Now I've tested the following scenarios:
- start vs code from start menu (not from a folder), File/Open file..., works (let's call this an "adhoc" file).
- start vs code from start menu (not from a folder), File/New file, set Lean4... works (let's call this an "untitled" file).
- start vs code from folder, open files in that folder, works!
   - now File/Open file... from somewhere else, also works!
   - now File/New file & set Lean4, also works!  
 
This last option now has 3 LeanClients running, one for the folder, and one for the adhoc package root, and one for 'untitled' files.  

For "adhoc" files the extension searches upwards in the directory hierarchy for a Lean package root and creates a new LeanClient for that root.

gebner commented on vscode-lean4/src/utils/clientProvider.ts:
```diff
@@ -103,29 +116,68 @@ export class LeanClientProvider implements Disposable {
+                // and we want that event callback to be the one that calls
+                // ensureClient, so we bail here so we don't try and do it twice
+                // on the same document.
+                return;
```
I vaguely remember that I didn't get the second didOpenEditor event when I tried it last time.  (That's why the condition was fall-through.)

lovettchris replied:
seems reliable now, but we can keep an eye on if, if you find a repro for when the second open event does not happen let me know...

gebner commented on vscode-lean4/src/utils/clientProvider.ts:
```diff
@@ -156,12 +219,10 @@ export class LeanClientProvider implements Disposable {
                 // ignore workspaces that belong to a different version of Lean.
                 console.log(`Lean4 extension ignoring workspace '${folderUri}' because it is not a Lean 4 workspace.`);
                 this.pending.delete(path);
+                this.clients.delete(path);
```
Should we dispose the client object in this case?

lovettchris replied:
Good catch, fixed.

gebner commented:
> In this case should we walk up the directory structure and find the Leanpkg.lean in lib\lean\src

If the user opens that `IO.lean` file directly, then this would be the ideal behavior.  (Maybe look for something else than `Leanpkg.lean` though, because that file no longer exists in recent nightlies.)

When the user opens that file via go-to-definition, it would also be reasonable to associate it with the lean server from the other file.  An argument for reusing servers is that find-references only works across the same server.

lovettchris commented:
> Maybe look for something else than `Leanpkg.lean` though

Right I meant to say I would be looking for the usual package files `leanpkg.toml` or `lean-toolchain`..

lovettchris commented:
@gebner ok, I did a big cleanup on when and how we walk directory looking for Lean package roots, they are now stateless helper functions.  This also cleaned up some redundant calls to testLeanVersion that happen on first Lean extension activation - so the first lean file now gets to running LSP state a bit faster than before. I also implemented the adhoc directory walk which seems to be working nicely.  For example, I do an adhoc open of a file deep in mathlib4 repo and it correctly finds the math4 root and opens only one LeanClient for all of mathlib4 which is the right thing to do.  So I think this PR is in good state now and ready to merge.

gebner commented on vscode-lean4/src/utils/projectInfo.ts:
```diff
@@ -0,0 +1,146 @@
+    if (packageFileUri.path.endsWith(tomlFileName))
+    {
+        return new Promise<string>((resolve, reject) => {
+            if (fs.existsSync(url)) {
+                fs.readFile(url, { encoding: 'utf-8' }, (err, data) =>{
```
Please stick to either async or sync.  Reading a small local file is fast enough that sync would be okay, but mixing them is not really helpful (since it could still block).  The `exists` operation is also deprecated (and shouldn't be used together with readFile due to the race condition where the file is deleted between the exists and readFile calls).

So either:
```typescript
try {
  fs.readFileSync(url, {encoding: 'utf-8'}).toString()
  // ...
} catch (err) {
  // could check for err.code === 'ENOENT'
  return null;
}
```

Or async, which is just as easy:
```typescript
try {
  (await fs.promises.readFile(url, {encoding: 'utf-8'})).toString()
  // ...
} catch (err) {
  // could check for err.code === 'ENOENT'
  return null;
}
```

I would also try to drop the duplicate `existsSync` check in `readLeanVersion`.

lovettchris replied:
Generally I agree with you but the only reason I didn't use exists() here is because it is deprecated for some reason.  In general I prefer all IO to be async as hard drives and network drives can have random very long delays depending on what's going on and I hate apps that force me to wait for them to do synchronous IO... but if you feel strongly about it I can change this to readFileSync - it will have ripple effects all over removing async behavior up the chain.

gebner commented on vscode-lean4/src/utils/projectInfo.ts:
```diff
@@ -0,0 +1,146 @@
+
+// Find the root of a Lean project and return the Uri for the package root and the Uri
+// for the 'leanpkg.toml' or 'lean-toolchain' file found there.
+export function findLeanPackageRoot(uri: Uri) : [WorkspaceFolder, Uri,Uri] {
+    if (!uri) return [null, null, null];
```
```suggestion
export function findLeanPackageRoot(uri: Uri) : [WorkspaceFolder | null, Uri | null, Uri | null] {
    if (!uri) return [null, null, null];
```

I'm surprised that this passes the type checker.  Another more self-explanatory return type might be `{workspaceFolder?: WorkspaceFolder, leanpkgToml?: Uri, leanToolchain?: Uri}`.
